crates/driver/src/tests/cases/settle.rs

```diff
@@ -110,3 +118,164 @@ async fn high_gas_limit() {
         .unwrap();
     test.settle(&id).await.ok().await;
 }
+
+#[tokio::test]
+#[ignore]
+async fn discards_excess_settle_and_solve_requests() {
+    let test = Arc::new(
+        tests::setup()
+            .allow_multiple_solve_requests()
+            .pool(ab_pool())
+            .order(ab_order())
+            .solution(ab_solution())
+            .settle_submission_deadline(6)
+            .done()
+            .await,
+    );
+
+    // MAX_SOLUTION_STORAGE = 5. Since this is hardcoded, no more solutions can be
+    // stored.
+    let solution_ids = join_all(vec![
+        test.solve(),
+        test.solve(),
+        test.solve(),
+        test.solve(),
+        test.solve(),
+    ])
+    .await
+    .into_iter()
+    .map(|res| res.ok().id())
+    .collect::<Vec<_>>();
+
+    let unique_solutions_count = solution_ids
+        .clone()
+        .into_iter()
+        .collect::<HashSet<_>>()
+        .len();
+    assert_eq!(unique_solutions_count, solution_ids.len());
+
+    // Disable auto mining to accumulate all the settlement requests.
+    test.disable_auto_mining().await;
+
+    // `collect_vec` is required to receive results in the same order.
+    let settlements = {
+        let test_clone = Arc::clone(&test);
+        solution_ids
+            .into_iter()
+            .map(|id| {
+                let test_clone = Arc::clone(&test_clone);
+                async move { test_clone.settle(&id).await }
+            })
+            .collect_vec()
+    };
+    let results_fut = tokio::spawn(join_all(settlements));
+
+    tokio::time::sleep(Duration::from_secs(2)).await;
+    // While there is no room in the settlement queue, `/solve` requests must be
+    // rejected.
+    test.solve().await.err().kind("TooManyPendingSettlements");
+
+    // Enable auto mining to process all the settlement requests.
+    test.enable_auto_mining().await;
+
+    let results = results_fut.await.unwrap();
+    assert_eq!(results.len(), 5);
+
+    // The first settlement must be successful.
+    results[0].clone().ok().await.ab_order_executed(&test).await;
+
+    let err_kinds = results[1..]
+        .iter()
+        .cloned()
+        .counts_by(|settle| settle.err().get_kind());
+    let queue_is_full_err_count = *err_kinds.get("TooManyPendingSettlements").unwrap();
+    let failed_to_submit_err_count = *err_kinds.get("FailedToSubmit").unwrap();
+    // There is a high possibility of a race condition where the first settlement
+    // request is dequeued at an unpredictable time. Therefore, we can't guarantee
+    // the order of the errors received, but their count must be persistent.
+    assert!(queue_is_full_err_count == 2 && failed_to_submit_err_count == 2);
```

Comment:
The request queue gets flushed whenever a request exceeds the related submission deadline, right?
Isn't that enough to make this more predictable? I'd like to better understand why we apparently can't fix this race condition altogether.

Reply:
Not sure how this is going to help here. But I found a way how to avoid race conditions:
1. Submit the first settlement. Sleep for a bit to make sure it gets dequeued and it's execution is paused.
2. Submit all the remaining settle requests.